ocaml/xapi/xapi_host.ml

```diff
@@ -2904,14 +2905,90 @@ let apply_updates ~__context ~self ~hash =
   in
   Db.Host.set_last_software_update ~__context ~self
     ~value:(get_servertime ~__context ~host:self) ;
-  Repository.apply_immediate_guidances ~__context ~host:self ~guidances ;
-  (* The warnings may not be returned to async caller if this happens
-   * on the coordinator host and the 'restartToolstack' is required.
-   *)
-  warnings
+  List.map
+    (fun g ->
+      [
+        Api_errors.updates_require_recommended_guidance
+      ; Updateinfo.Guidance.to_string g
+      ]
+    )
+    guidances
+  @ warnings
 
 let set_https_only ~__context ~self ~value =
   let state = match value with true -> "close" | false -> "open" in
   ignore
   @@ Helpers.call_script !Xapi_globs.firewall_port_config_script [state; "80"] ;
   Db.Host.set_https_only ~__context ~self ~value
+
+let try_restart_device_models_for_recommended_guidances ~__context ~host =
+  (* Restart device models of all running HVM VMs on the host by doing
+   * local migrations if it is required by recommended guidances. *)
+  Repository_helpers.do_with_device_models ~__context ~host
+  @@ fun (ref, record) ->
+  if
+    List.mem `restart_device_model
+      (Db.VM.get_recommended_guidances ~__context ~self:ref)
+  then
+    match
+      ( record.API.vM_power_state
+      , Helpers.has_qemu_currently ~__context ~self:ref
+      )
+    with
+    | `Running, true ->
+        Xapi_vm_migrate.pool_migrate ~__context ~vm:ref ~host
+          ~options:[("live", "true")] ;
+        None
+    | `Paused, true ->
+        error "VM 'ref=%s' is paused, can't restart device models for it"
+          (Ref.string_of ref) ;
+        Some ref
+    | _ ->
+        (* No device models are running for this VM *)
+        None
+  else
+    None
+
+let apply_recommended_guidances ~__context ~self =
+  try
+    let open Updateinfo in
+    Db.Host.get_recommended_guidances ~__context ~self |> function
+    | [] ->
+        try_restart_device_models_for_recommended_guidances ~__context
+          ~host:self
+    | [`reboot_host] ->
+        reboot ~__context ~host:self
+    | [`restart_toolstack] ->
+        try_restart_device_models_for_recommended_guidances ~__context
+          ~host:self ;
+        restart_agent ~__context ~host:self
+    | l ->
+        let host' = Ref.string_of self in
+        error
+          "Found wrong guidance(s) when applying recommended guidances on host \
+           ref='%s': %s"
+          host'
+          (String.concat ";"
+             (List.map Guidance.to_string
+                (List.map
+                   (fun ug ->
+                     match ug with
+                     | `reboot_host ->
+                         Guidance.RebootHost
+                     | `reboot_host_on_livepatch_failure ->
+                         Guidance.RebootHostOnLivePatchFailure
+                     | `restart_toolstack ->
+                         Guidance.RestartToolstack
+                     | `restart_device_model ->
+                         Guidance.RestartDeviceModel
```

Comment:
Please make this an independent function and place it somewhere where it can be reused (maybe in repository.ml?)
The you can use that function here, instead of having so many nesting levels here

Reply:
Added a function "of_update_guidance" in module Guidance, which transforms a "update_guidance" to Guidance.t.